prow/metrics/prowjobs/collector.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prowjobs
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/tools/cache"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+type prowJobLifecycleCollector struct {
+	HistogramVec *prometheus.HistogramVec
+}
+
+func (pjc *prowJobLifecycleCollector) Update(oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) {
+	if oldJob == nil {
+		return
+	}
+	if oldJob.Status.State != newJob.Status.State {
+
+		labels := getJobLabel(oldJob, newJob)
+		histogram, err := pjc.HistogramVec.GetMetricWithLabelValues(labels.values()...)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to get a histogram for a prowjob")
+			return
+		}
+
+		var oldTime metav1.Time
```

Comment:
nit: you could simplify this logic by storing the pointers


```go
var oldTime, newTime *metav1.Time

switch oldJob.Status.State {
case prowapi.TriggeredState:
    oldTime = &oldJob.CreationTimestamp
case prowapi.PendingState:
    oldTime = oldJob.PendingTimestamp
}
switch newJob.Status.State {
case prowapi.FailureState, prowapi.SuccessState, prowapi.ErrorState, prowapi.AbortedState:
    oldTime = newJob.CompletionTimestamp
case prowapi.PendingState:
    oldTime = newJob.PendingTimestamp
}
if oldTime == nil || newTime == nil {
    return
}
```

Reply:
Done.